Clear disk cache if it grows too large (dev_hdd1/cache)

MSuih opened the pull request:
The folder dev_hdd1/cache is a temporary folder and is used by games to temporarily store data outside of main memory. On PS3, this folder limited to 1 GB (afaik) and the console will automatically delete ~~files out of it to keep it under that treshold~~ everything inside it when switching games.

However, on this emulator the folder is never cleared and thus can grow to tens of gigabytes over time. While it is not a huge issue, I still believe that the emulator should keep the size of this folder in check so that it woudn't rob too much disc space. After all, these are just temporary files and thus aren't interesting enough to store permanently.

This PR checks the cache size and deletes olderst folders from it untill it's less than the treshold. Users can disable the feature and customize the delete threshold if they wish so.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -1797,6 +1799,31 @@ int main_window::IsValidFile(const QMimeData& md, QStringList* dropPaths)
 	return dropType;
 }
 
+void main_window::CheckCacheSize()
```
this doesnt belong into gui code

Megamouse commented on rpcs3/rpcs3qt/settings_dialog.cpp:
```diff
@@ -783,6 +783,15 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> guiSettings, std:
 
 	xemu_settings->EnhanceCheckBox(ui->enableHostRoot, emu_settings::EnableHostRoot);
 	SubscribeTooltip(ui->enableHostRoot, json_sys["enableHostRoot"].toString());
+	xemu_settings->EnhanceCheckBox(ui->enableCacheClearing, emu_settings::LimitCacheSize);
```
no newline

AniLeo commented:
This partition is always wiped clean when you boot a new game. Better to just clean everything inside it on boot or something.

Megamouse commented on rpcs3/rpcs3qt/emu_settings.cpp:
```diff
@@ -318,7 +318,7 @@ void emu_settings::EnhanceCheckBox(QCheckBox* checkbox, SettingsType type)
 }
 
-void emu_settings::EnhanceSlider(QSlider* slider, SettingsType type)
+void emu_settings::EnhanceSlider(QSlider* slider, SettingsType type, int multiplier)
```
why use this weird multiplier. why not just use mb with a propper stepping.

MSuih replied:
I disagree, but I guess you're the boss. I'll set it to range of 10k and tick marks every 1024 I guess.

Megamouse replied:
I just don't see why I should be limited to 10 values when I can use all 10240 values in MB.
Also the max size of a cache isnt exactly something you need to pick exactly

MSuih commented:
Hmm, that'd simplify the code a lot. Is anyone against clearing the cache on every boot? I was worried that it'd cause some games to write stuff there on every boot and thus drain write cycle unneccessarily.

kd-11 commented:
Yes, very much so. Some games (like DeS) can take minutes to boot without a cache, but less than 5 seconds with one available. Maybe users with the games on an SSD aren't affected as much but it can be very measurable. If the cache size is done per-game, you can lower the storage significantly - even 256M is enough imo.

MSuih commented:
Detecting what folders games use is a nightmare though, since games seem to be able to name them freely.  For example my cache has folders `BCJS-30015` (hypen, and with a product ID I do not have), `FROSTBITE` and `RATCHET`. How would I know which one belongs to what game?

kd-11 commented:
I don't know how much of it is cleared on game boot, especially since games seem to explicitly check for files in hdd1/cache on initial boot. It can be simplified to wipe on boot, but I reckon limiting the size is a good option for those willing to sacrifice some disk space for faster loading times. Or only clear if the booting game is not the same as the last game played, although I'm not sure how feasible this is.

Megamouse commented:
The last played game is actually saved in the gui settings.

MSuih commented:
I was under the impression that you want to separate this from the gui. Besides, I think it'd be useful to keep the cache intact in case someone wants to switch between games without having their cache wiped clean every time. I prefer my initial implementation as of now, but please do continue the discussion.

Kravickas commented:
Yeah what I was thinkink before reading, slider for amounth of GB u want to keep 0-?? and check it on start up

MSuih commented:
My current plan is to iterate through the `dev_hdd1/cache` folder, mapping out the last modification date and size of each top level folder (or file). Once that's complete I sort them by modification date and start deleting oldest folders untill the cache shrinks down to 80% of the cap.

How's that sound? The APis don't provide sizes and modification information on folders, so I'd have to recursively iterate through them to find everything, but hopefully that isn't dreadfully slow. I'll probably change the default size to 3 GB as well, it should mean that there's less to delete each time it's triggered.

Marthgun commented:
just a thought, aside from perpetual iteration you could store folder data, like the creation date somewhere, I know there's a cmd line function in windows you can use to get it, but Idk if you'd want to do it that way, i do think it might be the fastest though.

MSuih commented:
Yes, I think ill head for something like that instead (and there's functionality for checking basic information about file or folder). I'll think more about it when I have time later this week.

MSuih commented:
* Moved logic to System.cpp.
* Switched all instances of disc to disk.
* New deletion logic removes oldest folders based on their modification date (last time files were added/deleted/renamed/moved) and removes files until only 80% of the limiter is remaining (to avoid running the deletion on every startup if a game adds couple of bytes to folder).
* Changed the default size to 3GB, which is still more than what PS3 provides.
* Did the requested UI changes.

kd-11 commented on rpcs3/Emu/System.cpp:
```diff
@@ -494,8 +495,75 @@ bool Emulator::BootRsxCapture(const std::string& path)
+		return;
+	}
+
+	//retrieve items to delete
```
In this file, all the comments are missing a space after the double slash.

kd-11 commented on rpcs3/Emu/System.cpp:
```diff
@@ -494,8 +495,75 @@ bool Emulator::BootRsxCapture(const std::string& path)
+			break;
+	}
+
+	LOG_SUCCESS(GENERAL, "Cleaned disk cache, removed %f MB", size / 1024.0 / 1024.0);
```
Consider better formatting hint, such as %3.2f to avoid getting ridiculous IEEE representaions. Its not much use to get a fraction with 6 decimals, or worse, an exponent representation.

elad335 commented on rpcs3/Emu/System.cpp:
```diff
@@ -494,8 +495,83 @@ bool Emulator::BootRsxCapture(const std::string& path)
+	{
+		const std::string &name = std::get<0>(file);
+		if (fs::is_dir(name))
+			fs::remove_all(name, true);
```
put braces inbetween if and else blocks.

MSuih replied:
I don't really like how it makes the code look, but fine.

Megamouse commented:
you can still add bigger single step and page step values to the slider if you want.
This is doable in the Qt Designer.

MSuih commented:
I set page step to 512 units. Single step can stay at 1 so that people who want to be punctual can pick an exact value.

Nekotekina commented on rpcs3/Emu/System.h:
```diff
@@ -404,6 +407,9 @@ struct cfg_root : cfg::node
 		cfg::_bool host_root{this, "Enable /host_root/"};
 		cfg::_bool init_dirs{this, "Initialize Directories", true};
 
+		cfg::_bool limit_cache_size{this, "Limit disk cache size", true};
```
Default to false.

Nekotekina commented on rpcs3/Emu/System.cpp:
```diff
@@ -494,8 +495,87 @@ bool Emulator::BootRsxCapture(const std::string& path)
+	}
+
+	LOG_SUCCESS(GENERAL, "Cleaning disk cache...");
+	std::forward_list<std::tuple<std::string, u64, s64>> file_list{}; // name, size, last modified
```
No reason to use forward_list here. Tuple is also a bad choice. Could be a vector of fs::dir_entry, the type accessible by fs::dir iterator.

fargostz commented:
Is there a way to add the option to wype the cache or not so that way people can choose either way they want :)

RainbowCookie32 commented:
> Is there a way to add the option to wype the cache or not so that way people can choose either to delete the content or not :)

That's already there, under config->System you can select whether you want it to be cleaned automatically or not

fargostz commented:
> 
> 
> > Is there a way to add the option to wype the cache or not so that way people can choose either to delete the content or not :)
> 
> That's already there, under config->System you can select whether you want it to be cleaned automatically or not

Didn't see it my bad !

MSuih commented:
Neko wanted this to be disabled by default, so you need to go to the options menu if you want to enable this feature.
